### PR TITLE
Avoid certain cross-activity-infections

### PR DESCRIPTION
**Pull request by tschlenther**
in the snz scenario, agents can perform activities of different types in the same episimcontainer.
we might want to forbid infection for certain activity type pairs such as home <-> shopping.

when different activity types are allowed, the defined contact intentisities can differ.

### CODE REVIEW

File: src/main/java/org/matsim/episim/model/DefaultInfectionModel.java [hunk 1]
```diff
@@ -61,6 +69,27 @@ private void infectionDynamicsGeneralized(EpisimPerson personLeavingContainer, E
...
+            if (infectionSituation.equals(InfectionSituation.Facility)){
+                //home can only interact with home or leisure
+                if (infectionType.contains("home") && ! infectionType.contains("leis")  && ! ( leavingPersonsActivity.contains("home") && otherPersonsActivity.contains("home")  )){
+                    continue;
+                } else if (infectionType.contains("edu") && ! infectionType.contains("work") && ! ( leavingPersonsActivity.contains("edu") && otherPersonsActivity.contains("edu") )){
+                    //edu can only interact with work or edu
+                    continue;
+                }
+
+                trackContactPerson(personLeavingContainer, otherPerson, leavingPersonsActivity);
```
Comment (rakow) [inline src/main/java/org/matsim/episim/model/DefaultInfectionModel.java]: You changed the order of this block, which is now before `isRelevantForInfecttionDynamics`. Thus, the results are different and the tests failing.

Comment (tschlenther) [reply src/main/java/org/matsim/episim/model/DefaultInfectionModel.java]: The question is: which persons should be tracked? Probably that should be all persons that somebody is in contact with, i.e. contact persons that fullfill the conditions on activity types and are not in quarantine or hospital. The check whether the interaction is relevant for infection dynamic can be processed later.

I implemented the additional check, whether a person is traceable in the following commit.

In my opinion, the order is correct now. But please review and comment :)

**Comment by tschlenther** [comment]
The integration test failures do result from the changed order. I looked at input and output files.
Basically, the spreading dynamic is slightly slower (2-3 iterations difference at 90-100 iterations) but has the same results concercing susceptible and recovered people in the end.

So i will fix the test by changing the input.

**Comment by rakow** [comment]
`isRelevantForInfecttionDynamics` also applies the restrictions for certain acitivities. I think you now track every person even if they would not have performed this activitiy. i.e. the persons never met and there should be no tracking in this case I think.

**Comment by tschlenther** [comment]
> 
> 
> `isRelevantForInfecttionDynamics` also applies the restrictions for certain acitivities. I think you now track every person even if they would not have performed this activitiy. i.e. the persons never met and there should be no tracking in this case I think.

yes you are right. 
i changed it now to:

1) check whether leaving person can have interactions
2) draw contact person
3) check whether contact person can have interactions
4) check whether leaving person and contact person can have interaction with each other
5) execute tracking
6) check whether both have the same disease status
7) perform infection model

File: src/main/java/org/matsim/episim/EpisimUtils.java [hunk 2]
```diff
@@ -29,4 +29,27 @@ public static boolean hasStatusRelevantForInfectionDynamics(EpisimPerson personW
...
+            case infectedButNotContagious:
+                return true;
+            case contagious:
+                return true;
+            case seriouslySick:
+                return false; // assume is in hospital
+            case critical:
+                return false; // assume is in hospital
+            case recovered:
+                return true;
```
Comment (rakow) [inline src/main/java/org/matsim/episim/EpisimUtils.java]: Recovered persons will not be tracked, because they are filtered earlier.

**Comment by tschlenther** [comment]
@smuellervsp and i concluded:
The logic should actually be:

1. Is leaving person actually on the go?

start loop

2. draw possible contact person (increment counter)
3. is contact person actually on the go?
4. can they interact with each other? (cross activity type check)
5. tracking
6. can these persons infect each other (disease statuses check)
8. infection model

File: src/main/java/org/matsim/episim/EpisimUtils.java (hunk 2 above)
Comment (tschlenther) [reply src/main/java/org/matsim/episim/EpisimUtils.java]: we now decided to split isRelevantForInfectionDynamics up again into 2 checks:
1) is the person actually on the go?
2) has the person a relevant disease status for infection dynamics?

That is to account for conditions for tracking (people have contact with each other) separately from conditions for infection (disease status must 'fit')

**Pull request merged**